Implementation of MailboxConfiguration.

VaughnVernon opened the pull request:
@bwehrle Please check to ensure this is what we discussed. I have only created the interfaces and implementations for `MailboxConfiguration`. After you confirm I will add the abstraction over plugin to the `World` and tests.

VaughnVernon commented:
@bwehrle The implementation needs strengthening since currently the `name` must be the first property to be set. I will first collect all values and then generate the `Properties` on request.

bwehrle reviewed (changes requested):
This looks like what we discussed- am I correct that if I use a configuration that the name must be unique?  Or if I request an actor with a mailbox configuration a new mailbox is always created?  I'm not clear if its the name of the mailbox configuration, or the name of the mailbox, and what happens if there are 2 requests with the same mailboxconfiguration.name  value..

VaughnVernon commented:
@bwehrle Thanks for the feedback. I changed some attribute names to clarify the interfaces:

- The `name` was changed to `mailboxName`
  - It is indeed the name of the mailbox that will be registered with the `World`
    - Each configuration instance is registered only once and the mailbox type by the `mailboxName` is available all the time that the `World` is available
  - You will use this `mailboxName` to request that a new instance of the mailbox of that name be allocated for a given actor
- I changed the mailbox type names to:
  - `ArrayQueueConfiguration`
  - `ConcurrentQueueConfiguration`
  - `SharedRingBufferConfiguration`
- You may now request an instance of any of the three configuration types from a MailboxConfiguration factory by the name
  - `ArrayQueueConfiguration arrayQueueConfiguration()`
  - `ConcurrentQueueConfiguration concurrentQueueConfiguration()`
  - `SharedRingBufferConfiguration sharedRingBufferConfiguration()`

bwehrle commented:
This sounds perfect, thanks for making those clarifications.  I think this will really make it easy for the developer to create and use custom mailbox configuration.  Thank you @VaughnVernon

bwehrle reviewed (approved):
Thanks for making the requested minor changes.  Looks great!

VaughnVernon commented:
@bwehrle This should be the full implementation. Please see the tests and confirm and I will merge (or maybe you can?).

VaughnVernon commented:
@bwehrle Please see latest and full implementation. If you approve we will include in the next release, which is as soon as you confirm or request changes. (We want to release cluster elasticity improvements.)

jakzal commented on src/main/java/io/vlingo/xoom/actors/plugin/mailbox/agronampscarrayqueue/ManyToOneConcurrentArrayQueuePlugin.java:
```diff
@@ -26,6 +26,10 @@ public ManyToOneConcurrentArrayQueuePlugin() {
     this.configuration = new ManyToOneConcurrentArrayQueuePluginConfiguration();
   }
 
+  public ManyToOneConcurrentArrayQueuePlugin(PluginConfiguration configuration) {
```
Not `final` anymore?

VaughnVernon replied:
@jakzal Oh, I just missed one or some. I am out but can change later, unless you want to.

jakzal replied:
That's what I thought. Just pushed the fix.

VaughnVernon replied:
👍🏻 Thanks!
